#7815 Delayed synchronization

BatPio opened the pull request:
Fixes #7815
Fixes #2323 

<!--
TESTING

Writing tests is very important. Please try to write some tests for your PR. 
If you need help, please do not hesitate to ask in this PR for help.

Unit tests: https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#unit-tests
Instrumented tests: https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#instrumented-tests
UI tests: https://github.com/nextcloud/android/blob/master/CONTRIBUTING.md#ui-tests
 -->
- [ ] Tests written, or not not needed
Some tests are written. I'm waiting for suggestions on what else can be tested.

Delay configuration:

![sync_delay](https://user-images.githubusercontent.com/23415958/211904737-f2b55dbf-bbbb-495e-9bc0-e358c882341d.png)

BatPio committed:
# Conflicts:
#	CHANGELOG.md
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/66.json
#	app/src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java
#	app/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt
#	app/src/main/res/values-b+en+001/strings.xml
#	app/src/main/res/values-bg-rBG/strings.xml
#	app/src/main/res/values-cs-rCZ/strings.xml
#	app/src/main/res/values-da/strings.xml
#	app/src/main/res/values-de/strings.xml
#	app/src/main/res/values-es/strings.xml
#	app/src/main/res/values-fr/strings.xml
#	app/src/main/res/values-hu-rHU/strings.xml
#	app/src/main/res/values-pl/strings.xml
#	app/src/main/res/values-tr/strings.xml
#	app/src/main/res/values-zh-rCN/strings.xml
#	app/src/main/res/values-zh-rHK/strings.xml
#	app/src/main/res/values-zh-rTW/strings.xml
#	app/src/main/res/values/strings.xml

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
# Conflicts:
#	CHANGELOG.md
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/66.json
#	app/src/main/java/com/owncloud/android/ui/preview/PreviewMediaFragment.java
#	app/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt
#	app/src/main/res/values-b+en+001/strings.xml
#	app/src/main/res/values-bg-rBG/strings.xml
#	app/src/main/res/values-cs-rCZ/strings.xml
#	app/src/main/res/values-da/strings.xml
#	app/src/main/res/values-de/strings.xml
#	app/src/main/res/values-es/strings.xml
#	app/src/main/res/values-fr/strings.xml
#	app/src/main/res/values-hu-rHU/strings.xml
#	app/src/main/res/values-pl/strings.xml
#	app/src/main/res/values-tr/strings.xml
#	app/src/main/res/values-zh-rCN/strings.xml
#	app/src/main/res/values-zh-rHK/strings.xml
#	app/src/main/res/values-zh-rTW/strings.xml
#	app/src/main/res/values/strings.xml

Signed-off-by: batpio <prbator@gmail.com>

# Conflicts:
#	app/src/androidTest/java/com/owncloud/android/AbstractIT.java
#	app/src/main/java/com/nextcloud/client/jobs/FilesSyncWork.kt
#	app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java
#	app/src/main/java/com/owncloud/android/ui/activity/SyncedFoldersActivity.kt
#	app/src/main/java/com/owncloud/android/ui/dialog/SyncedFolderPreferencesDialogFragment.java
#	app/src/main/java/com/owncloud/android/ui/dialog/parcel/SyncedFolderParcelable.java

BatPio committed:
Delayed synchronization

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Delayed synchronization

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Licences

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Licences

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Tuning

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Tuning

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Tuning

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Tuning

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Rename .java to .kt

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Rename .java to .kt

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
New files converted to Kotlin

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
New files converted to Kotlin

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI hints fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI hints fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Lint fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Lint fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Lint fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Lint fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Screenshot test fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Screenshot test fixes

Signed-off-by: batpio <prbator@gmail.com>

AlvaroBrey commented on app/src/androidTest/java/com/owncloud/android/ui/dialog/DurationPickerDialogFragmentIT.kt:
```diff
@@ -0,0 +1,36 @@
+package com.owncloud.android.ui.dialog
```
Missing license. But thanks for writing tests!

BatPio replied:
Done, license added.

AlvaroBrey commented on app/src/androidTest/java/com/owncloud/android/utils/TimeUtilsTest.kt:
```diff
@@ -0,0 +1,28 @@
+package com.owncloud.android.utils
```
Missing license

BatPio replied:
Done, license added.

AlvaroBrey reviewed (changes requested):
I haven't tested this yet, especially to see how this will behave with existing autoupload folders. This can't be tested without adressing my first comment below, however.

Having said that, this looks quite promising! Thanks for contributing @BatPio

BatPio committed:
CR fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CR fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/68.json

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'nextcloud:master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge remote-tracking branch 'origin/feature/7815/delayed-synchronization-simplified-master-v2' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@AlvaroBrey Thanks for hints!  I made fixes

AlvaroBrey commented on app/src/main/java/com/nextcloud/client/database/entity/SyncedFolderEntity.kt:
```diff
@@ -54,6 +54,8 @@ data class SyncedFolderEntity(
     val uploadAction: Int?,
     @ColumnInfo(name = ProviderTableMeta.SYNCED_FOLDER_NAME_COLLISION_POLICY)
     val nameCollisionPolicy: Int?,
+    @ColumnInfo(name = ProviderTableMeta.SYNCED_FOLDER_UPLOAD_DELAY_TIME_MS)
+    val uploadDelayTimeMs: Int?,
```
```suggestion
    @ColumnInfo(name = ProviderTableMeta.SYNCED_FOLDER_UPLOAD_DELAY_TIME_MS, defaultValue = "0")
    val uploadDelayTimeMs: Long,
```
This does not need to be nullable, as the added code uses `cursor.getLong` (which coalesces `null` to 0 anyway). Additionally it's more correct to set this as a `Long`, even if the backing data type is `INTEGER` anyway, to prevent logical errors with  big numbers.

AlvaroBrey replied:
Note that the schema needs to be regenerated (by building the app again) if this is accepted

AlvaroBrey commented on app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java:
```diff
@@ -74,19 +74,22 @@ public void updateFilesystemFileAsSentForUpload(String path, String syncedFolder
     }
 
-    public Set<String> getFilesForUpload(String localPath, String syncedFolderId) {
+    public Set<String> getFilesForUpload(String localPath, String syncedFolderId, long uploadDelayTimeMs) {
```
Please rename the parameter to `minFileAge` or something like that, which is more appropriate for this method's scope.

Additionally you should check for `uploadDelayTimeMs > 0`, in order to prevent logical errors. In the case of `uploadDelayTimeMs <= 0`  just don't add the modified query.

BatPio replied:
I made fixes: https://github.com/nextcloud/android/pull/11278/files#diff-2663f36d409eaa2b3f0c5aaffe7a35ad03baaa0001ecf74677b19091f095ce6eR82 . 

I need to test the application after fixes. I'll let you know when I'm done.

BatPio replied:
@AlvaroBrey  Sorry for the delay, I was very busy last month. I've done and tested the fixes, immediate and delayed syncing seems to work fine

AlvaroBrey reviewed (changes requested):
Thanks @BatPio and sorry for the delay! This is a quality PR :) There's a couple of things I'd like you to change if possible, but after that, this is okay by me.

@tobiasKaminsky please review as well

BatPio committed:
Update app/src/main/java/com/nextcloud/client/database/entity/SyncedFolderEntity.kt

Co-authored-by: Álvaro Brey <alvaro.brv@gmail.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/java/com/nextcloud/client/database/entity/SyncedFolderEntity.kt

Co-authored-by: Álvaro Brey <alvaro.brv@gmail.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'nextcloud:master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio commented:
I need to test the application after fixes. I'll let you know when I'm done.

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
synchronization fix

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

andreapx commented:
@BatPio is this still a work in progress or has it been abandoned?

BatPio committed:
SyncedFolderPreferencesDialogFragment migration

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Rename .java to .kt

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
SyncedFolderPreferencesDialogFragment moved dialog

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2-update-to_master

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/69.json
#	app/screenshots/gplay/debug/com.nextcloud.client.SyncedFoldersActivityIT_testSyncedFolderDialog.png
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java
#	app/src/main/java/com/owncloud/android/ui/activity/SyncedFoldersActivity.kt
#	app/src/main/java/com/owncloud/android/ui/dialog/SyncedFolderPreferencesDialogFragment.kt
#	app/src/main/java/com/owncloud/android/ui/dialog/parcel/SyncedFolderParcelable.java

BatPio committed:
SyncedFolderPreferencesDialogFragment moved dialog

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Licence update

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Licence update

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
CI fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio commented:
@JonasMayerDev @alperozturk96 

Hi, sorry for delay, I updated my branch to master and now I have a problem with check:
"TooManyFunctions - 15/15 - [Class 'SyncedFolderPreferencesDialogFragment' with '15' functions detected. "

Can you give me some suggestions on what can be moved out of this class?

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2

alperozturk96 commented:
> @JonasMayerDev @alperozturk96
> 
> Hi, sorry for delay, I updated my branch to master and now I have a problem with check: "TooManyFunctions - 15/15 - [Class 'SyncedFolderPreferencesDialogFragment' with '15' functions detected. "
> 
> Can you give me some suggestions on what can be moved out of this class?

Hello

You can use

`@Suppress("TooManyFunctions")`

this annotation to suppress it.

BatPio committed:
@Suppress("TooManyFunctions")

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
@Suppress("TooManyFunctions")

Signed-off-by: batpio <prbator@gmail.com>

BatPio commented:
@alperozturk96 

Thanks for advice

I couldn't find any example of DialogManager in the project, so I used @Suppress("TooManyFunctions") . I think the presence of showUploadDelayDialog in SyncedFolderPreferencesDialogFragment is acceptable, because they are closely related.

I think MR is ready for CR

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2

uspecies commented:
i just wonder what the upper limit for the day value will be? as i guess i may choose everything between 1 to 12 months in days.
So for example 31 days may not satisfy all users #just saying :p 

i will start using NC after this PR is released :) cheers!

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/82.json
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java

BatPio committed:
Merged with main

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/84.json
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java

BatPio committed:
Merged with main

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merged with main

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

PhilLab commented:
This functionality is really helpful, thanks for investing the time here @BatPio. I was in the middle of setting up my dev environment to implement exactly this functionality when I saw that you already did the work 🙏 .

@tobiasKaminsky @alperozturk96  what is missing to get this merged?
1. Merge is blocked because changes were requested by @AlvaroBrey two years ago, but they seem to have been addressed by now. Can this be considered done?
2. Tests have been added. Are there more required?
3. Due to this PR being open a long time, there were a lot of merges from master. Do we need a cleanup of the git history on this branch (though non-trivial, due to all the merges from master) or can we leave it like it is?

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/src/main/java/com/nextcloud/client/jobs/FilesSyncWork.kt
#	app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java

BatPio committed:
Merged with main

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@PhilLab 
Right, let's finish the job! :)

@tobiasKaminsky @alperozturk96 
I've updated my branch to master and merged my mechanism with the latest changes in FilesSyncWork. My MR is ready for CR.

alperozturk96 reviewed (commented):
Thank you very much for the PR, and apologies for the delayed response. We strive to provide timely feedback, but at times, we need to handle other PRs, which can occasionally cause some to become stale.

Thank you again for your excellent work on this 💯. Before merging, I noticed a few minor things that need to be addressed. Additionally, I was unable to rebase to the master branch due to conflicts. Since these changes affect the auto-upload logic, we will need to conduct thorough testing before merging.


![Screenshot 2025-01-09 at 09 00 41](https://github.com/user-attachments/assets/32de04cd-58cc-44e9-9855-7e4eaf0bf613)

alperozturk96 reviewed (changes requested):
Explained above

BatPio committed:
Merge branch 'nextcloud:master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge remote-tracking branch 'origin/feature/7815/delayed-synchronization-simplified-master-v2' into feature/7815/delayed-synchronization-simplified-master-v2

tobiasKaminsky committed:
re-add lint check for translation

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

nextcloud-bot committed:
Fix(l10n): Update translations from Transifex

Signed-off-by: Nextcloud bot <bot@nextcloud.com>

nextcloud-bot committed:
Fix(l10n): Update translations from Transifex

Signed-off-by: Nextcloud bot <bot@nextcloud.com>

nextcloud-bot committed:
Fix(l10n): Update translations from Transifex

Signed-off-by: Nextcloud bot <bot@nextcloud.com>

tobiasKaminsky committed:
revert removing verification

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

nextcloud-bot committed:
Fix(l10n): Update translations from Transifex

Signed-off-by: Nextcloud bot <bot@nextcloud.com>

tobiasKaminsky committed:
bump lib

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

renovate[bot] committed:
fix(deps): update dependency com.android.tools.build:gradle to v8.8.0

Signed-off-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge remote-tracking branch 'origin/feature/7815/delayed-synchronization-simplified-master-v2' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Fixes after CR

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
DB automigration

Signed-off-by: batpio <prbator@gmail.com>

BatPio commented:
@alperozturk96 
Hi, thanks for the code review. I have made fixes.

I have rebased my branch to master. Next rebase should be easy, but there is a lot of commits in the history. I hope it is not a problem and they will disappear during merging with squash.

You can continue code review :)

BatPio commented on app/src/androidTest/java/com/owncloud/android/ui/dialog/DurationPickerDialogFragmentIT.kt:
```diff
@@ -0,0 +1,45 @@
+class DurationPickerDialogFragmentIT : AbstractIT() {
+
+    @get:Rule
+    val testActivityRule = IntentsTestRule(TestActivity::class.java, true, false)
```
fixed

BatPio commented on app/src/main/java/com/owncloud/android/ui/dialog/DurationPickerDialogFragment.kt:
```diff
@@ -0,0 +1,138 @@
+
+    private var resultListener: Listener? = null
+
+    private var _binding: DurationPickerBinding? = null
```
fixed

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@alperozturk96 

Unfortunately rebase still doesn't work on my branch. Can we update it by merge with commit and after review merge it with squash?

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

alperozturk96 commented:
@ZetaTom @tobiasKaminsky Could you review and test this PR?

alperozturk96 commented:
@BatPio Thank you so much for the update. Code looks ok to me but I want to test as well. 

How can I test the delay?

BatPio commented:
@alperozturk96 
Testing is time-consuming.

After setting the auto upload delay for a directory, you need to add a new file to that directory and wait. After the delay time has elapsed, the file should appear on the server. It is important to check that the file has not appeared on the server before the delay time has elapsed. Due to the Android's sync interval, the delay may increase by an additional 15 minutes.
In case of a 30-minute delay, the file should appear on server no earlier than after 30 minutes and no later than after 45 minutes.

I tested this with screenshots. Unfortunately, I have no idea how to test this without waiting.

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/86.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/nextcloud/client/jobs/FilesSyncWork.kt

BatPio committed:
Merge remote-tracking branch 'origin/feature/7815/delayed-synchronization-simplified-master-v2' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@alperozturk96 

I have adapted my code to the current solution with FilesystemDataProvider. MR is ready for testing

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

alperozturk96 commented:
@tobiasKaminsky @ZetaTom 

I tested auto upload seems like okay. Could you please also test, especially `FilesystemDataProvider.java `and `FilesSyncWork.kt` ?

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

ZetaTom reviewed (commented):
Thank you for taking the time to contribute these enhancements and for maintaining this pull request.

In general this seems to work. The duration selector dialogue is very intuitive. Great work!

The phrase *Delay file upload* suggests that the upload of a given file is delayed. However, we're only looking for files *older* than the specified time. This becomes an issue when there are files already present or when files are moved to the auto upload folder. These files will most likely be uploaded immediately. To address this issue, I would suggest to rename the feature to something like *Exclude recent files* or *Minimum file age*.

I think that we should involve @Hyeyoung346 in this discussion.

BatPio committed:
Update app/src/main/java/com/owncloud/android/ui/dialog/DurationPickerDialogFragment.kt

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/java/com/owncloud/android/ui/dialog/DurationPickerDialogFragment.kt

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/res/layout/duration_picker.xml

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/res/layout/duration_picker.xml

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/res/layout/duration_picker.xml

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Update app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java

Co-authored-by: Tom <70907959+ZetaTom@users.noreply.github.com>
Signed-off-by: BatPio <prbator@gmail.com>

BatPio committed:
Using kotlin Duration instead of TimeUtils

Signed-off-by: batpio <prbator@gmail.com>

BatPio commented on app/src/main/java/com/owncloud/android/utils/TimeUtils.kt:
Yes, you are right. I removed TimeUtils

BatPio commented on app/src/main/java/com/owncloud/android/ui/dialog/SyncedFolderPreferencesDialogFragment.kt:
```diff
@@ -569,5 +598,29 @@ class SyncedFolderPreferencesDialogFragment : DialogFragment(), Injectable {
         }
+
+        @Suppress("MagicNumber")
+        private fun getDelaySummary(context: Context, duration: Long): String {
```
I added spaces

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
DurationPicker layout fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
DurationPicker layout fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
DurationPicker labels

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
DurationPicker labels

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Renamed to Min file age

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/87.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt

BatPio committed:
Merge with master

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Renamed to Min file Age fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Renamed to Min file Age fixes

Signed-off-by: batpio <prbator@gmail.com>

jancborchardt commented:
When reading through the linked issues https://github.com/nextcloud/android/issues/7815 and https://github.com/nextcloud/android/issues/2323 – it seems that _if_ we implement this delay setting at all, then it should be only there when auto upload is set to "Delete after upload".
Also, instead of a flexible time picker, we should do it like elsewhere in Nextcloud and have some reasonable presets. E.g.:
- Upload immediately (default)
- 15 minutes
- 30 minutes
- 1 hour
- 24 hours
- 1 week

These presets already seem to cover all the cases brought up in the issues, without detailed time configuration needed.

"Send later" in Nextcloud Mail | "Clear status after"-picker in online status
-|-
![image](https://github.com/user-attachments/assets/93233183-23ff-4e1f-a984-1d7b231e99e4)|![image](https://github.com/user-attachments/assets/e8e5f896-f41d-4147-93eb-1bea200a69cb)



Do you see other solutions here, especially ones which don’t add another customizable setting but just do it automatically? @tobiasKaminsky @alperozturk96 @ZetaTom

jancborchardt commented:
@PhilLab ah, thanks for the note, good point.

Then it should just be changed to the suggested intervals as mentioned.

BatPio committed:
DurationPicker replaced with singleChoiceItem dialog

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@ZetaTom @jancborchardt 
Thank you for the code review, I made changes

@ZetaTom 
You are right, the phrase "Delay file upload" can be misleading. I changed it to "Minimum file age"

@jancborchardt 
I replaced DurationPicker with a list with fixed values

After combining both ideas:
![Screenshot_20250319_214900](https://github.com/user-attachments/assets/e29e4ece-d0b8-40ed-9ce9-8d3f156dce2d)![Screenshot_20250319_214929](https://github.com/user-attachments/assets/df750815-1fa7-4651-adaa-16e2989f809d)![Screenshot_20250319_214942](https://github.com/user-attachments/assets/80ed279f-89db-4ecf-8932-de5ebee8f909)

Since the sync interval in Android is 15 minutes, I think this value is not useful. Instead, I would add values of 6 hours and 3 days

jancborchardt reviewed (commented):
> Since the sync interval in Android is 15 minutes, I think this value is not useful. Instead, I would add values of 6 hours and 3 days

@BatPio sure, sounds good. :) Also, the wording of "No restrictions" could be changed to "Upload immediately" which might be a bit more clear?

BatPio committed:
Duration singleChoiceItem dialog values tuning

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

jancborchardt reviewed (approved):
Looks good from the design side now. :)

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio commented:
@alperozturk96 

I've reworked the UI as suggested by @jancborchardt . Could you please review the changes?

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/88.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt

BatPio committed:
Master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

jancborchardt commented:
@tobiasKaminsky @alperozturk96 @ZetaTom could you give this a review? :)

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

tobiasKaminsky commented:
Will review tomorrow

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/89.json

BatPio committed:
Master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-fixes

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/90.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt

BatPio committed:
Master merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-fixes

# Conflicts:
#	app/src/main/java/com/owncloud/android/ui/dialog/SyncedFolderPreferencesDialogFragment.kt

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-fixes

# Conflicts:
#	app/src/androidTest/java/com/nextcloud/client/SyncedFoldersActivityIT.java
#	app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java

BatPio committed:
ModificationDate fix

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2-utom-2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/91.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/nextcloud/client/jobs/FilesSyncWork.kt
#	app/src/main/java/com/owncloud/android/datamodel/FilesystemDataProvider.java
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java

BatPio committed:
Merge branch 'refs/heads/master' into feature/7815/delayed-synchronization-simplified-master-v2-utom-2

# Conflicts:
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java
#	app/src/main/java/com/owncloud/android/utils/FilesSyncHelper.java

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/schemas/com.nextcloud.client.database.NextcloudDatabase/97.json
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/owncloud/android/datamodel/SyncedFolderProvider.java
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>

BatPio committed:
Merge branch 'master' into feature/7815/delayed-synchronization-simplified-master-v2

# Conflicts:
#	app/src/main/java/com/nextcloud/client/database/NextcloudDatabase.kt
#	app/src/main/java/com/owncloud/android/db/ProviderMeta.java

BatPio committed:
Merge fixes

Signed-off-by: batpio <prbator@gmail.com>
